fms/utils/serialization.py

```diff
@@ -88,48 +107,117 @@ def get_adapted(
 from fms import models
 
 
+def get_ckp_format(model_path: Union[str, Path]) -> str:
+    """
+    Returns the checkpoint format of a model checkpoint. If format
+    is not recognized, assumes all files are regular pytorch
+    checkpoints and returns "pt"
+
+    Args:
+    model_path: the path to find the weights.
+    """
+    model_path = Path(os.path.expanduser(model_path))
+    if (
+        model_path.suffix == ".safetensors"
+        or len(sorted(model_path.glob("*.safetensors"))) > 0
+    ):
+        return "st"
+    if model_path.suffix == ".pth" or len(sorted(model_path.glob("*.pth"))) > 0:
+        return "pt"
+    if model_path.suffix == ".bin" or len(sorted(model_path.glob("*.bin"))) > 0:
+        return "hf"
+    return "pt_all"
+
+
+def get_safetensors_item(key, file: Path, device: torch.device) -> torch.Tensor:
+    from safetensors import safe_open  # type: ignore[import-untyped]
+
+    with torch.no_grad():
+        with safe_open(
+            file, framework="pt", device=str(device)
+        ) as model_weights:  # type: ignore[attr-defined]
+            return model_weights.get_tensor(key)
+
+
+class LazySafetensorsDict(collections.UserDict):
+    def set_lazy_tensor(self, key, file, device):
+        super().__setitem__(key, lambda: get_safetensors_item(key, file, device))
+
+    def __getitem__(self, key):
+        lazy_tensor = super().__getitem__(key)
+        if callable(lazy_tensor):
+            lazy_tensor = lazy_tensor()
+            super().__setitem__(key, lazy_tensor)
+        return lazy_tensor
+
+
 def load_state_dict(
     model_path: Union[str, Path],
+    *,
+    checkpoint_format: Optional[str] = None,
     distributed_strategy: Optional[str] = None,
     checkpoint_sharding: Optional[str] = None,
     initial_device: torch.device = torch.device("cpu"),
     rank: int = 0,
     world_size: int = 1,
-) -> Mapping[str, Any]:
+) -> MutableMapping[str, Any]:
     """
     Validates that the file(s) found at a checkpoint path are compatible with
-    the intended (possibly distributed) use-case, and returns the state dict
-    if needed.
+    the intended (possibly distributed) use-case, and returns a lazy loading
+    state dict if possible (some formats may not support that).
+
+    If model_path is a directory, it'll try to load, in this order, pytorch
+    models (.pth format), HF models (.bin format), and safetensors
+    (.safetensors format), unless checkpoint_format is specified.
 
     Args:
     model_path: the path to find the weights. If not set, return None.
+    checkpoint_format: how the checkpoint files are saved: None, 'pt',
+            'hf', or 'st'. If None, guess based on files.
```

Comment:
I think this description is a bit ambiguous since the "source" format is also "how the checkpoints are saved" in a sense. especially with pt_all, it sort of seems that maybe we really mean file suffix? If we did have this represent file suffix, then a user could pass in 'ckp' in the case of .ckp files that are actually pytorch file format... but file type could also make sense.

We could also guess the appropriate suffix based on what files are in the directory ... a user might have .ckp checkpoints and also some random .config / .param / .log files in the same directory. Right now there's no way to handle that, since a default glob of "*" would pick up arbitrary files too.

Or maybe we should be supporting something like `model_path='/some/directory/*.ckp'` ?

Reply:
I like supporting glob, and guessing the file format based on the source. The only issue is we can have e.g. both .bin and .safetensors coming from HF or .safetensors and .pth from meta sources. If you pass the glob in the path this is solved, but in case we just pass a folder, should we just associate an extension to each source?

Reply:
in the case of a directory without a glob, I think we'll just have to guess the desired suffix in some order of preferred file suffixes?